File: API/API/Controllers/ProjectController.vb
Imports System.Web.Http
Imports System.Net
Imports System.Collections.Generic

Namespace Controllers

    <RoutePrefix("api/projects")>
    <Authorize>
    Public Class ProjectController
        Inherits BaseApiController

        Private ReadOnly _projectService As IProjectService
        Private ReadOnly _taskService As ITaskService

        Public Sub New(projectService As IProjectService, taskService As ITaskService)
            _projectService = projectService
            _taskService = taskService
        End Sub

        <HttpGet>
        <Route("")>
        Public Function GetMyProjects() As IHttpActionResult
            Return Execute(Function() _projectService.GetProjectsForUser(CurrentUserId))
        End Function

        <HttpPost>
        <Route("")>
        Public Function Create(<FromBody> req As CreateProjectRequest) As IHttpActionResult
            If req Is Nothing Then
                Return Content(HttpStatusCode.BadRequest, New ApiResponse(Of Object)(False, "Dữ liệu không hợp lệ.", Nothing))
            End If
            Return Execute(Function() _projectService.CreateProject(req, CurrentUserId))
        End Function

        <HttpGet>
        <Route("{projectId:guid}")>
        <ProjectAuthorize>
        Public Function GetProjectDetail(projectId As Guid) As IHttpActionResult
            Return Execute(Function() _projectService.GetProjectDetail(projectId, CurrentUserId))
        End Function

        <HttpPut>
        <Route("{projectId:guid}")>
        <ProjectAuthorize("Owner")>
        Public Function UpdateProject(projectId As Guid, <FromBody> req As UpdateProjectRequest) As IHttpActionResult
            If req Is Nothing Then
                Return Content(HttpStatusCode.BadRequest, New ApiResponse(Of Object)(False, "Dữ liệu không hợp lệ.", Nothing))
            End If
            Return Execute(Function() _projectService.UpdateProject(projectId, req, CurrentUserId))
        End Function

        <HttpDelete>
        <Route("{projectId:guid}")>
        <ProjectAuthorize("Owner")>
        Public Function DeleteProject(projectId As Guid) As IHttpActionResult
            Return Execute(Sub() _projectService.DeleteProject(projectId, CurrentUserId))
        End Function

        ' ── MEMBERS MANAGEMENT ──────────────────────────

        <HttpGet>
        <Route("{projectId:guid}/members")>
        <ProjectAuthorize>
        Public Function GetMembers(projectId As Guid) As IHttpActionResult
            Return Execute(Function() _projectService.GetMembers(projectId))
        End Function

        <HttpPost>
        <Route("{projectId:guid}/members")>
        <ProjectAuthorize("Owner")>
        Public Function AddMember(projectId As Guid, <FromBody> req As AddMemberRequest) As IHttpActionResult
            If req Is Nothing Then
                Return Content(HttpStatusCode.BadRequest, New ApiResponse(Of Object)(False, "Dữ liệu không hợp lệ.", Nothing))
            End If
            Return Execute(Function() _projectService.AddMember(projectId, req, CurrentUserId))
        End Function

        <HttpPut>
        <Route("{projectId:guid}/members/{userId:guid}")>
        <ProjectAuthorize("Owner")>
        Public Function UpdateMemberRole(projectId As Guid, userId As Guid, <FromBody> req As UpdateMemberRequest) As IHttpActionResult
            If req Is Nothing Then
                Return Content(HttpStatusCode.BadRequest, New ApiResponse(Of Object)(False, "Dữ liệu không hợp lệ.", Nothing))
            End If
            Return Execute(Function() _projectService.UpdateMemberRole(projectId, userId, req, CurrentUserId))
        End Function

        <HttpDelete>
        <Route("{projectId:guid}/members/{userId:guid}")>
        <ProjectAuthorize("Owner")>
        Public Function RemoveMember(projectId As Guid, userId As Guid) As IHttpActionResult
            Return Execute(Sub() _projectService.RemoveMember(projectId, userId, CurrentUserId))
        End Function

        ' ── TASKS IN PROJECT ────────────────────────────

        <HttpGet>
        <Route("{projectId:guid}/tasks")>
        <ProjectAuthorize>
        Public Function GetProjectTasks(projectId As Guid) As IHttpActionResult
            Return Execute(Function() _taskService.GetProjectTasks(projectId, CurrentUserId))
        End Function

        <HttpPost>
        <Route("{projectId:guid}/tasks")>
        <ProjectAuthorize("Owner", "Editor")>
        Public Function CreateTask(projectId As Guid, <FromBody> req As CreateTaskRequest) As IHttpActionResult
            If req Is Nothing Then
                Return Content(HttpStatusCode.BadRequest, New ApiResponse(Of Object)(False, "Dữ liệu không hợp lệ.", Nothing))
            End If
            Return Execute(Function() _taskService.CreateTask(req, CurrentUserId, projectId))
        End Function

    End Class

End Namespace

File: API_v2/Controllers/ProjectController.cs
using System.Threading.Tasks;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using API_v2.Models.DTOs;
using API_v2.Services.Interfaces;

namespace API_v2.Controllers
{
    [Route("api/projects")]
    [Authorize]
    public class ProjectController : BaseApiController
    {
        private readonly IProjectService _projectService;

        public ProjectController(IProjectService projectService)
        {
            _projectService = projectService;
        }

        [HttpGet]
        public async Task<ActionResult> GetMyProjects([FromQuery] int page = 1, [FromQuery] int pageSize = 20)
        {
            page = Math.Max(page, 1);
            pageSize = Math.Clamp(pageSize, 1, 100);
            var result = await _projectService.GetProjectsForUserAsync(CurrentUserId, page, pageSize);
            return Ok(new ApiResponse<PagedResponse<ProjectResponse>>(true, "Success", result));
        }

        [HttpPost]
        [Authorize(Roles = "Admin,Manager")]
        public async Task<ActionResult> Create([FromBody] CreateProjectRequest req)
        {
            var result = await _projectService.CreateProjectAsync(req, CurrentUserId);
            return Ok(new ApiResponse<ProjectResponse>(true, "Project created successfully.", result));
        }

        [HttpGet("{projectId:guid}")]
        public async Task<ActionResult> GetProjectDetail(Guid projectId)
        {
            var result = await _projectService.GetProjectDetailAsync(projectId, CurrentUserId);
            return Ok(new ApiResponse<ProjectResponse>(true, "Success", result));
        }

        [HttpPut("{projectId:guid}")]
        public async Task<ActionResult> UpdateProject(Guid projectId, [FromBody] UpdateProjectRequest req)
        {
            var result = await _projectService.UpdateProjectAsync(projectId, req, CurrentUserId);
            return Ok(new ApiResponse<ProjectResponse>(true, "Project updated successfully.", result));
        }

        [HttpDelete("{projectId:guid}")]
        public async Task<ActionResult> DeleteProject(Guid projectId)
        {
            await _projectService.DeleteProjectAsync(projectId, CurrentUserId);
            return Ok(new ApiResponse<object>(true, "Project deleted successfully.", null));
        }

        [HttpGet("{projectId:guid}/members")]
        public async Task<ActionResult> GetMembers(Guid projectId)
        {
            var result = await _projectService.GetMembersAsync(projectId, CurrentUserId);
            return Ok(new ApiResponse<List<MemberResponse>>(true, "Success", result));
        }

        [HttpPost("{projectId:guid}/members")]
        public async Task<ActionResult> AddMember(Guid projectId, [FromBody] AddMemberRequest req)
        {
            var result = await _projectService.AddMemberAsync(projectId, req, CurrentUserId);
            return Ok(new ApiResponse<MemberResponse>(true, "Member added successfully.", result));
        }

        [HttpPut("{projectId:guid}/members/{userId:guid}")]
        public async Task<ActionResult> UpdateMemberRole(Guid projectId, Guid userId, [FromBody] UpdateMemberRequest req)
        {
            var result = await _projectService.UpdateMemberRoleAsync(projectId, userId, req, CurrentUserId);
            return Ok(new ApiResponse<MemberResponse>(true, "Member role updated successfully.", result));
        }

        [HttpDelete("{projectId:guid}/members/{userId:guid}")]
        public async Task<ActionResult> RemoveMember(Guid projectId, Guid userId)
        {
            await _projectService.RemoveMemberAsync(projectId, userId, CurrentUserId);
            return Ok(new ApiResponse<object>(true, "Member removed successfully.", null));
        }

    }
}
